spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/query/SpannerStatementQueryTests.java

```diff
@@ -0,0 +1,268 @@
+
+		Object[] params = new Object[] {
+				"BUY",
+				"abcd",
+				"abc123",
+		};
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void unsupportedParamTypeTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn(
+				"findTop3DistinctIdActionPriceByActionAndSymbolOrTraderIdAndPriceLessThanOrPriceGreater"
+						+ "ThanEqualAndIdIsNotNullAndTraderIdIsNullOrderByIdDesc");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[] {
+				"BUY",
+				"abcd",
+				"abc123",
+				8.88,
+				3.33,
+				new Trade(),
+				"ignored",
+		};
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test(expected = UnsupportedOperationException.class)
+	public void deleteTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn(
+				"deleteTop3DistinctIdActionPriceByActionAndSymbolOrTraderIdAndPriceLessThanOrPriceGreater"
+						+ "ThanEqualAndIdIsNotNullAndTraderIdIsNullOrderByIdDesc");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[0];
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test(expected = UnsupportedOperationException.class)
+	public void unSupportedPredicateTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn("countByTraderIdBetween");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[0];
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test
+	public void countTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn("countByAction");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[]{
+				"BUY",
+		};
+
+		List<Trade> results = new ArrayList<>();
+		results.add(new Trade());
+
+		when(spannerOperations.find(any(), (Statement) any(), any()))
+				.thenReturn((List) results);
+
+		assertEquals(1, partTreeSpannerQuery.execute(params));
+	}
+
+	@Test
+	public void existsTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn("existsByAction");
```

Comment:
same, confused by "byAction"

Reply:
please see explanation for countByAction